### PR TITLE
Fixed broken authentication in cadc

### LINKED ISSUE #1565: Authentication broken in CADC
The `cadc` package had a work-around for allowing users to send authenticated queries to the CADC TAP service. This solution is not working anymore with the release of `pyvo-1.0` which implements its own authentication scheme.

Might take the opportunity to also re-enable listing of jobs that is now supported by `pyvo`

### PR DESCRIPTION
**Pull request by andamian**
close #1565

### CODE REVIEW

**Comment by andamian** [comment]
Hmm! Have I've missed something from the change long? Otherwise, it's ready for review.

File: astroquery/cadc/core.py [hunk 1]
```diff
@@ -141,9 +153,12 @@ def login(self, user=None, password=None, certificate_file=None):
         certificates.
         """
         # start with a new session
-        pyvo.dal.tap.s = requests.Session()
+        if not isinstance(self.cadctap._session, authsession.AuthSession):
+            raise TypeError('Cannot login with user provided session that is '
+                            'not an pyvo.autsession.AuthSession')
```
Comment (cbanek) [inline astroquery/cadc/core.py:158]: missing an h in pyvo.autsession

Comment (cbanek) [reply astroquery/cadc/core.py:158]: so it was my hope that people could also pass in other classes here, like just a normal requests session, but I doubt this will come up any time soon.

File: astroquery/cadc/core.py [hunk 2]
```diff
@@ -428,7 +446,10 @@ def get_image_list(self, query_result, coordinates, radius):
                 '{}?{}'.format(self.data_link_url,
                                urlencode({'ID': pid_sublist}, True)))
             for service_def in datalink.bysemantics('#cutout'):
-                access_url = service_def.access_url.decode('ascii')
+                if isinstance(service_def.access_url, str):
+                    access_url = service_def.access_url
+                else:
+                    access_url = service_def.access_url.decode('ascii')
```
Comment (cbanek) [inline astroquery/cadc/core.py:452]: this change seems unrelated?

File: astroquery/cadc/core.py (hunk 1 above)
Comment (cbanek) [reply astroquery/cadc/core.py:158]: Ah, I think I see, because you take the client certificate, you want it to be an authsession so you can call set_client_ceritificate, which only exists on that.  But for client certificates, it's not really a login anyway right?  Maybe just have client certificate use a different call than login?

I think another option might be just to set it to a fresh new authsession no matter what, after all, you're doing a login, so maybe that implies a logout of what was before?  Like in logout you're clearing everything (which I think is the right thing to do).

Anyway, nitpicking.

File: .travis.yml [hunk 3]
```diff
@@ -115,7 +115,7 @@ jobs:
           stage: Test docs, astropy dev, and without optional dependencies
           env: PYTHON_VERSION=3.8
                CONDA_DEPENDENCIES='requests beautifulsoup4 html5lib keyring'
-               PIP_DEPENDENCIES='https://github.com/keflavich/httpbin/archive/master.zip pytest-astropy'
+               PIP_DEPENDENCIES='https://github.com/keflavich/httpbin/archive/master.zip pytest-astropy pyvo'
```
Comment (bsipocz) [inline .travis.yml:118]: This travis job is for having only the bare minimum, mandatory dependencies. `pyvo` is still "optional" only, so please do not add it here.

A separate question whether we switch it to be a requirement. We said we would do that once the gaia module moved on to use it, too, however I don't know any ETA for that, so maybe we should just make that move without waiting any longer. I'll open a separate issue for the discussion.

File: astroquery/cadc/core.py [hunk 4]
```diff
@@ -160,8 +175,8 @@ def login(self, user=None, password=None, certificate_file=None):
                 "Content-type": "application/x-www-form-urlencoded",
                 "Accept": "text/plain"
             }
-            response = pyvo.dal.tap.s.post(login_url, data=args,
-                                           headers=header)
+            response = requests.post(login_url, data=args,
```
Comment (bsipocz) [inline astroquery/cadc/core.py:178]: any reason why not go through `self._request` instead?

File: astroquery/cadc/tests/DummyTapHandler.py [hunk 5]
```diff
@@ -27,9 +27,9 @@ def check_method(self, method):
         if method == self.__invokedMethod:
             return
         else:
-            raise Exception("Method '"+str(method) +
-                            "' not invoked. (Invoked method is '" +
-                            str(self.__invokedMethod)+"')")
+            raise Exception("Method '"+str(method)
+                            + "' not invoked. (Invoked method is '"
```
Comment (bsipocz) [inline astroquery/cadc/tests/DummyTapHandler.py:31]: no need for the `+`, and it would be more cleaner to use an fstring instead.

File: astroquery/cadc/core.py (hunk 1 above)
Comment (andamian) [reply astroquery/cadc/core.py:158]: `login` was here from the beginning (historical reasons). In fact it should be called `authenticate` and should allow multiple logins to the same service (cert, cookie etc).
I think I've tried it with a session of type `requests.Session` and it works.
Truth is I can probably login when the session is a simple `request.Session` which I'll change it to do. The difference is that such a simple session does not know to match credentials with the type of end point so that would be a reason for people to use 'AuthSession' instead.

File: astroquery/cadc/core.py (hunk 2 above)
Comment (andamian) [reply astroquery/cadc/core.py:452]: Indeed. It's related to a change in votable. Breaks the integration tests right now.

File: .travis.yml (hunk 3 above)
Comment (andamian) [reply .travis.yml:118]: Hmm. I was under the impression that mandatory dependencies are listed in `setup.py`. To run the cadc tests I've had `pyvo` specified in `PIP_DEPENDENCIES` before. And it's still there for Python other than version 3.8. Hence my update.

File: astroquery/cadc/core.py (hunk 4 above)
Comment (andamian) [reply astroquery/cadc/core.py:178]: Forgot about it. Sorry. I'll make the correction.

File: .travis.yml (hunk 3 above)
Comment (bsipocz) [reply .travis.yml:118]: yes, but in this job we only install the mandatory ones to make sure we indeed don't always rely on optional ones. Naturally it means that for this job no cadc tests are run due to the pyvo dependency.

File: astroquery/cadc/core.py (hunk 1 above)
Comment (andamian) [reply astroquery/cadc/core.py]: Updated a bunch more. Made it so it also uses user credentials with a `requests.Session`. Hopefully it's more clearer now. This is until we need to put it all together and have A&A working with other related services such as `Datalink` and `SODA`. It should be good for now.

File: .travis.yml (hunk 3 above)
Comment (andamian) [reply .travis.yml]: They are both called `PIP_DEPENDENCIES` hence the confusion. I've removed it from P3.8 now.

File: astroquery/cadc/tests/DummyTapHandler.py (hunk 5 above)
Comment (andamian) [reply astroquery/cadc/tests/DummyTapHandler.py]: It turns out it was old dead code. Removed.

File: astroquery/cadc/core.py (hunk 4 above)
Comment (andamian) [reply astroquery/cadc/core.py]: Fixed although I couldn't make use of the bonus functionality (cache) as the call retrieves time sensitive cookies.

File: astroquery/cadc/core.py [hunk 6]
```diff
@@ -89,7 +91,11 @@ def __init__(self, url=None, tap_plus_handler=None, verbose=None):
             a url to use instead of the default
         tap_plus_handler : deprecated
         verbose : deprecated
-
+        session: `requests.Session` type
```
Comment (bsipocz) [inline astroquery/cadc/core.py:94]:
```suggestion
        auth_session: `requests.Session` type
```

File: astroquery/cadc/core.py [hunk 7]
```diff
@@ -135,19 +153,29 @@ def login(self, user=None, password=None, certificate_file=None):
...
+            raise TypeError('Cannot login with user provided session that is '
+                            'not an pyvo.authsession.AuthSession or '
+                            'requests.Session')
+        if not certificate_file and not (user and password):
             raise AttributeError('login credentials missing (user/password '
                                  'or certificate)')
-        else:
+        if certificate_file:
+            if isinstance(self.cadctap._session, authsession.AuthSession):
+                self.cadctap._session.credentials.\
```
Comment (bsipocz) [inline astroquery/cadc/core.py:168]: we change the suggested line lenght in astropy to 100, so this should be fine. (no need to change now, I'm saying more as an info)

**Comment by bsipocz** [comment]
@keflavich or @ceb8 - I leave this open for a little while in case you also want to have a look, otherwise I'm happy with merging.

File: astroquery/cadc/core.py (hunk 6 above)
Comment (andamian) [reply astroquery/cadc/core.py:94]: How do you do this? I would like to update type too to include `pyvo.auth.authsession.AuthSession` without pushing from local repo.

Comment (andamian) [reply astroquery/cadc/core.py]: Done. Thanks

**Comment by bsipocz** [comment]
travis failure is a red herring and due to a missing conda feedstock merge, which will be hopefully resolved by tomorrow.

**Comment by bsipocz** [comment]
thanks @andamian!
